pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,327 @@
+}
+
+// ProxyRelayMeter is the off-chain Supplier's rate limiter.
+// It ensures that no application is over-serviced by the Supplier by maintaining
+// the max amount of stake the supplier can consume per session and the amount of
+// stake consumed by mined relays.
+type ProxyRelayMeter struct {
+	// The known applications that have their stakes metered.
+	apps map[string]*appRelayMeter
+	// overServicingAllowance adjusts the max amount to allow for controlled over-servicing
+	// or, if negative, to create a stake buffer.
+	// TODO_TECHDEBT(@red-0ne): Expose overServicingAllowance as a configuration parameter.
+	overServicingAllowance int64
+
+	applicationQuerier client.ApplicationQueryClient
+	serviceQuerier     client.ServiceQueryClient
+	sharedQuerier      client.SharedQueryClient
+	eventsQueryClient  client.EventsQueryClient
+	blockQuerier       client.BlockClient
+
+	relayMeterMu sync.Mutex
+}
+
+func NewRelayMeter(deps depinject.Config) (relayer.RelayMeter, error) {
+	rm := &ProxyRelayMeter{
+		apps:                   make(map[string]*appRelayMeter),
+		overServicingAllowance: 1000000,
+	}
+
+	if err := depinject.Inject(
+		deps,
+		&rm.sharedQuerier,
+		&rm.applicationQuerier,
+		&rm.serviceQuerier,
+		&rm.blockQuerier,
+		&rm.eventsQueryClient,
+	); err != nil {
+		return nil, err
+	}
+
+	return rm, nil
+}
+
+// Start starts the relay meter by observing application staked events and new sessions.
+func (rmtr *ProxyRelayMeter) Start(ctx context.Context) error {
+	eventsObs, err := rmtr.eventsQueryClient.EventsBytes(ctx, "tm.event = 'Tx'")
+	if err != nil {
+		return err
+	}
+
+	// Listen to application staked events and update known application stakes.
+	appStakedEvents := filterTypedEvents[*apptypes.EventApplicationStaked](ctx, eventsObs, nil)
+	channel.ForEach(ctx, appStakedEvents, rmtr.forEachEventApplicationStakedFn)
+
+	// Listen to new blocks and reset the relay meter application stakes every new session.
+	committedBlocksSequence := rmtr.blockQuerier.CommittedBlocksSequence(ctx)
+	channel.ForEach(ctx, committedBlocksSequence, rmtr.forEachNewBlockFn)
+
+	return nil
+}
+
+// ClaimRelayPrice claims the relay price for the given relay request metadata.
+// It deducts the relay cost from the application's stake and returns an error if
+// the application has been rate limited.
+func (rmtr *ProxyRelayMeter) ClaimRelayPrice(ctx context.Context, reqMeta servicetypes.RelayRequestMetadata) error {
+	rmtr.relayMeterMu.Lock()
+	defer rmtr.relayMeterMu.Unlock()
+
+	sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
+	if err != nil {
+		return err
+	}
+
+	service, err := rmtr.serviceQuerier.GetService(ctx, reqMeta.SessionHeader.ServiceId)
+	if err != nil {
+		return err
+	}
+
+	serviceRelayDifficulty, err := rmtr.serviceQuerier.GetServiceRelayDifficulty(ctx, service.Id)
+	if err != nil {
+		return err
+	}
+
+	appAddress := reqMeta.SessionHeader.ApplicationAddress
+
+	appMetrics, ok := rmtr.apps[appAddress]
```

Comment:
This line is confusing. I'd expect one of the following (or other) as an example:

```go
app, ok := rmtr.apps[appAddr]
appRelayMeter, ok := rmtr.appRelayMeters[appAddr]
appMetrics, ok := rmtr.appMetricsMap[appAddr]
```